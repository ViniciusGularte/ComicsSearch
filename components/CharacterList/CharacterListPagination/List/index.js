import React from "react";
import Link from "next/link";
import {
  Container,
  CharacterCard,
  BannerImage,
  CharacterCardTextContent,
  CardTextName,
  CardTextFooter,
} from "./styles";
const List = ({ characters, only_favorite, stateCharacters }) => {
  return (
    <Container className="characters-list">
      {characters &&
        characters.results &&
        characters.results.length > 0 &&
        characters.results.map((character, i) => {
          return (
            <CharacterCard className="characters" key={i}>
              <BannerImage
                background={
                  only_favorite ? character.image : character.image.medium_url
                }
              />
              <CharacterCardTextContent>
                <CardTextName>
                  {stateCharacters[character.id] &&
                  stateCharacters[character.id].name
                    ? stateCharacters[character.id].name
                    : character.name}{" "}
                  <hr />
                </CardTextName>
                <CardTextFooter>
                  <Link href={`/character/${encodeURIComponent(character.id)}`}>
                    See More
                  </Link>
                </CardTextFooter>
              </CharacterCardTextContent>
            </CharacterCard>
          );
        })}
    </Container>
  );
};
export default List;
