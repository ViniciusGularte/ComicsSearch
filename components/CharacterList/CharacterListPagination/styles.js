import styled from "styled-components";

const LoadingItem = styled.div`
  display: flex;
  flex: 1;
  margin: 16px;
  text-align: center;
  justify-content: center;
  align-items: center;
  flex-direction: row;
  justify-content: space-between;
  font-size: 18px;
  color: ${(props) => props.theme.primary};
`;
const ContainerSearchView = styled.div`
  flex: 1;
  justify-content: center;
  display: flex;
  flex-direction: column;
`;
const ContainerSearchViewText = styled.div`
  flex: 1;
  align-self: center;
  font-size: 16px;
  font-weight: bold;
`;
const ContainerSearchViewButton = styled.button`
  font-size: 16px;
  border: none;
  background: none;
  cursor: pointer;
  margin: 0;
  padding: 0;
  flex: 1;
  color: ${(props) => props.theme.primary};
`;
export {
  ContainerSearchView,
  ContainerSearchViewText,
  ContainerSearchViewButton,
  LoadingItem,
};
