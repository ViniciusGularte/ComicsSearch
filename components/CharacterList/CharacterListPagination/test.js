import React from "react";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import CharacterListPagination from "./index";
const initialState = {
  characters: {},
};
const emptyCharacterResponse = {
  error: "OK",
  offset: 0,
  number_of_total_results: 50,
  results: [],
};

Enzyme.configure({ adapter: new Adapter() });

describe("CharacterListPagination", () => {
  const mockStore = configureStore();
  let store;
  it("no favorite | no search", () => {
    store = mockStore(initialState);
    const wrapper = shallow(
      <Provider store={store}>
        <CharacterListPagination
          is_search={false}
          only_favorite={false}
          Itens={emptyCharacterResponse}
        />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
  it("with favorite | no search", () => {
    store = mockStore(initialState);
    const wrapper = shallow(
      <Provider store={store}>
        <CharacterListPagination
          is_search={false}
          only_favorite={true}
          Itens={emptyCharacterResponse}
        />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
  it("no favorite | with search", () => {
    store = mockStore(initialState);
    const wrapper = shallow(
      <Provider store={store}>
        <CharacterListPagination
          is_search={true}
          only_favorite={true}
          Itens={emptyCharacterResponse}
        />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
