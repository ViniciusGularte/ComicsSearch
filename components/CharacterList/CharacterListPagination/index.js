import React, { useEffect, useState } from "react";
import Router, { useRouter } from "next/router";
import {
  LoadingItem,
  ContainerSearchView,
  ContainerSearchViewText,
  ContainerSearchViewButton,
} from "./styles";
import List from "./List/index";
import { useSelector } from "react-redux";
import { ContainerError } from "../../../assets/globalStyles/styles";
const CharacterList = ({ Itens, is_search, only_favorite }) => {
  const [characters, setCharacters] = useState({
    results: [],
  });
  const [viewSearchResults, setViewSearchResults] = useState(false);
  const [toogleSearchResults, setToogleSearchResults] = useState(false);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(false);

  const setLoadingTrue = () => setLoading(true);
  const setLoadingFalse = () => setLoading(false);
  const router = useRouter();
  const stateCharacters = useSelector((state) => state.characters);
  const toogleViewSearchResults = () => {
    setViewSearchResults(false);
  };
  useEffect(() => {
    if (Itens) {
      if (only_favorite) {
        const array_characters_store = Object.values(stateCharacters);
        const array_characters_favorite = array_characters_store.filter(
          (character) => character.is_favorite
        );
        setToogleSearchResults(true);
        setCharacters({
          results: array_characters_favorite,
        });
      } else if (Itens.error !== "OK") {
        setError(true);
      } else {
        setError(false);
        if (is_search) {
          setViewSearchResults(true);
          setToogleSearchResults(true);
          setCharacters({
            results: Itens.results,
            offset: Itens.offset,
            number_of_total_results: Itens.number_of_total_results,
          });
        } else {
          setCharacters({
            ...characters,
            results: toogleSearchResults
              ? Itens.results
              : [...characters.results, ...Itens.results],
            offset: Itens.offset,
            number_of_total_results: Itens.number_of_total_results,
          });
          setToogleSearchResults(false);
        }
      }
    }
  }, [Itens]);

  useEffect(() => {
    window.addEventListener("scroll", handleScroll);
    return () => {
      window.removeEventListener("scroll", handleScroll);
    };
  });
  useEffect(() => {
    Router.events.on("routeChangeStart", setLoadingTrue);
    Router.events.on("routeChangeComplete", setLoadingFalse);
    return () => {
      Router.events.off("routeChangeStart", setLoadingTrue);
      Router.events.off("routeChangeComplete", setLoadingFalse);
    };
  }, []);

  const handleScroll = () => {
    const lastCharacter = document.querySelector(
      ".characters-list > .characters:last-child"
    );
    if (lastCharacter) {
      const lastUserLoadedOffset =
        lastCharacter.offsetTop + lastCharacter.clientHeight;
      const pageOffset = window.pageYOffset + window.innerHeight;
      if (pageOffset > lastUserLoadedOffset) {
        if (
          characters.offset < characters.number_of_total_results &&
          !loading
        ) {
          const query = router.query;
          query.offset = parseInt(characters.offset) + 50;
          router.push({
            pathname: router.pathname,
            query: query,
          });
        }
      }
    }
  };
  return (
    <>
      {error ? (
        <ContainerError>Error</ContainerError>
      ) : viewSearchResults && is_search ? (
        <ContainerSearchView>
          <ContainerSearchViewText>
            Founded {characters.results.length} characters
          </ContainerSearchViewText>
          {characters.results.length > 0 ? (
            <ContainerSearchViewButton onClick={toogleViewSearchResults}>
              Click here to view
            </ContainerSearchViewButton>
          ) : null}
        </ContainerSearchView>
      ) : (
        <List
          characters={characters}
          only_favorite={only_favorite}
          stateCharacters={stateCharacters}
        />
      )}
      {loading && !is_search && !only_favorite && (
        <LoadingItem> Loading ... </LoadingItem>
      )}
    </>
  );
};
export default CharacterList;
