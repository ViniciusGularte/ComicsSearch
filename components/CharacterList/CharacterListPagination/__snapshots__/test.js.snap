// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterListPagination no favorite | no search 1`] = `ShallowWrapper {}`;

exports[`CharacterListPagination no favorite | with search 1`] = `ShallowWrapper {}`;

exports[`CharacterListPagination with favorite | no search 1`] = `ShallowWrapper {}`;
